aquadoggo/src/domain.rs

```diff
@@ -0,0 +1,1060 @@
+                None => Ok(SeqNum::default()),
+            }?;
+
+            // Check if skiplink is required and if it is get the entry and return it's hash
+            let skiplink = if is_lipmaa_required(seq_num.as_u64()) {
+                // Determine skiplink ("lipmaa"-link) entry in this log.
+                Some(get_expected_skiplink(store, public_key, &log_id, &seq_num).await?)
+            } else {
+                None
+            }
+            .map(|entry| entry.hash());
+
+            next_args.backlink = latest_entry.map(|entry| entry.hash().into());
+            next_args.skiplink = skiplink.map(|hash| hash.into());
+            next_args.seq_num = seq_num.into();
+            next_args.log_id = log_id.into();
+        }
+    };
+
+    Ok(next_args)
+}
+
+/// Persist an entry and operation to storage after performing validation of claimed values against
+/// expected values retrieved from storage.
+///
+/// Returns the arguments required for constructing the next entry in a bamboo log for the specified
+/// author and document.
+///
+/// This method is intended to be used behind a public API and so we assume all passed values
+/// are in themselves valid.
+///
+/// The steps and validation checks this method performs are:
+///
+/// Validate the values encoded on entry against what we expect based on our existing stored entries:
+/// - verify the claimed sequence number against the expected next sequence number for the author and log
+/// - get the expected backlink from storage
+/// - get the expected skiplink from storage
+/// - verify the bamboo entry (requires the expected backlink and skiplink to do this)
+/// Ensure single node per author
+/// - @TODO
+/// Validate operation against it's claimed schema
+/// - @TODO
+/// Determine document id
+/// - if this is a create operation
+///   - derive the document id from the entry hash
+/// - in all other cases
+///   - verify that all operations in previous_operations exist in the database
+///   - verify that all operations in previous_operations are from the same document
+///   - ensure the document is not deleted
+/// - verify the claimed log id matches the expected log id for this author and log
+/// Determine the next arguments
+/// Persist data
+/// - if this is a new document
+///   - store the new log
+/// - store the entry
+/// - store the opertion
+/// Return next entry arguments
+pub async fn publish<S: StorageProvider>(
+    store: &S,
+    entry_encoded: &EntrySigned,
+    operation_encoded: &OperationEncoded,
+) -> Result<NextEntryArguments> {
+    ////////////////////////////////
+    // DECODE ENTRY AND OPERATION //
+    ////////////////////////////////
+
+    let entry = decode_entry(entry_encoded, Some(operation_encoded))?;
+    let operation = Operation::from(operation_encoded);
+    let author = entry_encoded.author();
+    let log_id = entry.log_id();
+    let seq_num = entry.seq_num();
+
+    ///////////////////////////
+    // VALIDATE ENTRY VALUES //
```

Comment:
Could be cool to move this into a separate function, just as it is somehow its own "logical" flow?

Reply:
Yeh, could do, the `latest_entry`/`backlink` is used later on in the returned arg so need to keep that accessible, but should be possible.

Reply:
I had a look at this but I'd rather leave it for now, I see that it is a logical "unit" but refactoring nicely would add in an extra db call and as we don't need this pattern anywhere elsewhere _yet_ I'd rather leave it for now.